File: packages/prompt-core-python/src/agentmark/prompt_core/base_adapter.py
"""Base adapter primitives shared by all SDK-specific adapters.

Mirrors `@agentmark-ai/prompt-core/base-adapter`. Absorbs the three
responsibilities that every adapter was doing by hand:

  1. Holding a registry of in-process tools + an McpServerRegistry.
  2. Resolving `tools: [...]` lists — including `mcp://server/*` wildcards.
  3. Building telemetry metadata.

SDK adapters subclass `BaseAdapter[TTool]`, pass in an MCP client factory,
and implement their SDK-native `adapt_text` / `adapt_object` / etc by
calling `resolve_tools()` + `apply_param_map()` with an SDK-specific map.
"""

from __future__ import annotations

import json
from collections.abc import Callable
from dataclasses import dataclass
from typing import Any, Generic, TypeVar

from .mcp import McpServers, parse_mcp_uri
from .mcp_registry import McpClientFactory, McpServerRegistry

TTool = TypeVar("TTool")


@dataclass
class ParamMapTransform:
    """Entry that renames AND transforms a value, e.g. wrapping a number
    in an SDK-specific helper. Used when a plain rename isn't enough."""

    key: str
    transform: Callable[[Any], Any]


# A ParamMap entry is one of:
#   - str: rename (input key → output key), pass value through
#   - None: drop the field entirely
#   - ParamMapTransform: rename + transform value
ParamMapEntry = str | None | ParamMapTransform
ParamMap = dict[str, ParamMapEntry]


def apply_param_map(
    input_obj: dict[str, Any] | None, param_map: ParamMap
) -> dict[str, Any]:
    """Translate a snake_case settings dict to SDK-native params.

    Unknown keys (not present in `param_map`) are dropped silently — adapters
    opt fields in explicitly so the wire format stays predictable.
    """
    if not input_obj:
        return {}
    out: dict[str, Any] = {}
    for in_key, value in input_obj.items():
        if value is None:
            continue
        entry = param_map.get(in_key)
        if entry is None or not isinstance(entry, (str, ParamMapTransform)):
            # Drop: entry missing OR explicitly `None`
            continue
        if isinstance(entry, str):
            out[entry] = value
        else:
            out[entry.key] = entry.transform(value)
    return out


def build_telemetry_metadata(
    telemetry: dict[str, Any] | None,
    props: dict[str, Any] | None,
    prompt_name: str,
    agentmark_meta: dict[str, Any] | None = None,
) -> dict[str, Any] | None:
    """Merge user telemetry with prompt-level metadata. Shared across
    adapters so fixes land everywhere at once."""
    if telemetry is None:
        return None
    existing_meta = telemetry.get("metadata") or {}
    metadata = {
        **existing_meta,
        "prompt_name": prompt_name,
        "props": json.dumps(props or {}),
    }
    if agentmark_meta:
        metadata.update(agentmark_meta)
    return {**telemetry, "metadata": metadata}


class BaseAdapter(Generic[TTool]):
    """Base class every SDK adapter extends.

    Subclasses provide the SDK-native tool type via the TTool generic and
    wire in the MCP client factory at construction. Beyond that, subclasses
    only implement `adapt_text` / `adapt_object` / etc by calling
    `resolve_tools()` + `apply_param_map()`.
    """

    def __init__(
        self,
        mcp_client_factory: McpClientFactory[TTool],
        tools: dict[str, TTool] | None = None,
        mcp_servers: McpServers | None = None,
    ) -> None:
        self._tools: dict[str, TTool] | None = tools
        self._mcp_registry: McpServerRegistry[TTool] = McpServerRegistry(
            mcp_client_factory
        )
        if mcp_servers:
            self._mcp_registry.register_servers(mcp_servers)

    @property
    def mcp_registry(self) -> McpServerRegistry[TTool]:
        return self._mcp_registry

    async def resolve_tools(self, tool_names: list[str]) -> dict[str, TTool]:
        """Resolve a list of tool refs. Supports plain names,
        `mcp://server/tool`, and `mcp://server/*` wildcard expansion."""
        out: dict[str, TTool] = {}
        for tool_name in tool_names:
            if tool_name.startswith("mcp://"):
                parsed = parse_mcp_uri(tool_name)
                server, tool = parsed["server"], parsed["tool"]
                if tool == "*":
                    all_tools = await self._mcp_registry.get_all_tools(server)
                    out.update(all_tools)
                    continue
                out[tool] = await self._mcp_registry.get_tool(server, tool)
                continue
            if self._tools is not None and tool_name in self._tools:
                out[tool_name] = self._tools[tool_name]
                continue
            available = (
                ", ".join(self._tools.keys()) if self._tools else "(none)"
            )
            raise ValueError(
                f"Tool '{tool_name}' referenced in prompt config was not "
                f"found in the provided tools record. Available tools: {available}"
            )
        return out

File: packages/prompt-core/src/base-adapter.ts
import type { AdaptOptions } from "./types";
import type { McpServers } from "./mcp";
import { parseMcpUri } from "./mcp";
import { McpServerRegistry, type McpClientFactory } from "./mcp-registry";

/**
 * Declarative field-rename entry for translating snake_case AgentMark config
 * to SDK-native camelCase params.
 *
 * - `string`: rename `input[inputKey]` → `output[string]`, pass value through.
 * - `null`: drop this field entirely (e.g. field isn't supported by this SDK).
 * - function: rename + transform value. Used for cases like
 *   `max_calls` → `stopWhen: stepCountIs(n)` where the target SDK needs
 *   a wrapped value rather than the raw number.
 */
export type ParamMapEntry =
  | string
  | null
  | { key: string; transform: (value: any) => unknown };

export type ParamMap = Record<string, ParamMapEntry>;

/**
 * Translate a snake_case settings object to SDK-native params using the
 * declarative map. Unknown keys are dropped silently — adapters opt fields
 * in explicitly so the wire format stays predictable.
 */
export function applyParamMap(
  input: Record<string, unknown> | undefined,
  map: ParamMap
): Record<string, unknown> {
  if (!input) return {};
  const out: Record<string, unknown> = {};
  for (const [inKey, value] of Object.entries(input)) {
    if (value === undefined) continue;
    const entry = map[inKey];
    if (entry === undefined || entry === null) continue;
    if (typeof entry === "string") {
      out[entry] = value;
      continue;
    }
    out[entry.key] = entry.transform(value);
  }
  return out;
}

/**
 * Shared telemetry metadata builder. Every adapter merges the user's base
 * telemetry with the prompt name, props (stringified), and optional
 * agentmark_meta block. Keeping this in one place means fixes to telemetry
 * encoding land everywhere at once.
 */
export function buildTelemetryMetadata(
  telemetry: AdaptOptions["telemetry"],
  props: Record<string, unknown>,
  promptName: string,
  agentmarkMeta?: Record<string, unknown>
) {
  if (!telemetry) return undefined;
  return {
    ...telemetry,
    metadata: {
      ...telemetry?.metadata,
      prompt_name: promptName,
      props: JSON.stringify(props),
      ...(agentmarkMeta ?? {}),
    },
  };
}

/**
 * Base class every SDK adapter extends. Absorbs the three responsibilities
 * that were duplicated across ai-sdk-v4, ai-sdk-v5, and mastra adapters:
 *
 *   1. Holding a registry of in-process tools + an McpServerRegistry.
 *   2. Resolving `tools: [...]` lists — including `mcp://server/*` wildcards.
 *   3. Building telemetry metadata.
 *
 * Subclasses provide the SDK-native tool type via `TTool` and wire in the
 * MCP client factory in their constructor. Beyond that, subclasses only
 * implement `adaptText` / `adaptObject` / `adaptImage` / `adaptSpeech`
 * by calling `resolveTools()` + `applyParamMap()` with their SDK's paramMap.
 */
export abstract class BaseAdapter<TTool = unknown> {
  protected readonly tools: Record<string, TTool> | undefined;
  protected readonly mcpRegistry: McpServerRegistry<TTool>;

  constructor(
    mcpClientFactory: McpClientFactory<TTool>,
    tools?: Record<string, TTool>,
    mcpServers?: McpServers
  ) {
    this.tools = tools;
    this.mcpRegistry = new McpServerRegistry<TTool>(mcpClientFactory);
    if (mcpServers) {
      this.mcpRegistry.registerServers(mcpServers);
    }
  }

  /**
   * Resolve a list of tool references from prompt config. Supports:
   *   - Plain names: must exist in the adapter's `tools` record.
   *   - `mcp://server/tool`: single tool from MCP server.
   *   - `mcp://server/*`: all tools from MCP server (wildcard expansion).
   */
  protected async resolveTools(
    toolNames: string[]
  ): Promise<Record<string, TTool>> {
    const out: Record<string, TTool> = {};
    for (const toolName of toolNames) {
      if (toolName.startsWith("mcp://")) {
        const { server, tool } = parseMcpUri(toolName);
        if (tool === "*") {
          const all = await this.mcpRegistry.getAllTools(server);
          Object.assign(out, all);
          continue;
        }
        out[tool] = await this.mcpRegistry.getTool(server, tool);
        continue;
      }
      if (this.tools && toolName in this.tools) {
        out[toolName] = this.tools[toolName];
        continue;
      }
      const available = this.tools
        ? Object.keys(this.tools).join(", ")
        : "(none)";
      throw new Error(
        `Tool '${toolName}' referenced in prompt config was not found in the provided tools record. Available tools: ${available}`
      );
    }
    return out;
  }
}
